### PR TITLE
Merchant dashboard

### LINKED ISSUE #29: User Story 42, Merchant's Items index page
```
As a merchant
When I visit my dashboard
I see a link to view my own items
When I click that link
My URI route should be "/dashboard/items"
```

### Checklist:

- [ ] base tests are written
- [ ] base code is written to pass tests
- [ ] edge case ("sad path") tests written (if required)
- [ ] edge case code written (if required)
- [ ] code is reviewed
- [ ] branch is merged
- [ ] all tests still pass

child of #33

child of #79

depends on #57

### LINKED ISSUE #30: User Story 41, Merchant Dashboard Statistics
```
As a merchant
When I visit my dashboard, I see an area with statistics:
- top 5 items I have sold by quantity, and the quantity of each that I've sold
- total quantity of items I've sold, and as a percentage against my sold units plus remaining inventory (eg, if I have sold 1,000 things and still have 9,000 things in inventory, the message would say something like "Sold 1,000 items, which is 10% of your total inventory")
- top 3 states where my items were shipped, and their quantities
- top 3 city/state where my items were shipped, and their quantities (Springfield, MI should not be grouped with Springfield, CO)
- name of the user with the most orders from me (pick one if there's a tie), and number of orders
- name of the user who bought the most total items from me (pick one if there's a tie), and the total quantity
- top 3 users who have spent the most money on my items, and the total amount they've spent
```

### Checklist:

- [ ] base tests are written
- [ ] base code is written to pass tests
- [ ] edge case ("sad path") tests written (if required)
- [ ] edge case code written (if required)
- [ ] code is reviewed
- [ ] branch is merged
- [ ] all tests still pass

child of #33

child of #79

depends on #32

### LINKED ISSUE #31: User Story 40, Merchant Dashboard displays Orders
```
As a merchant
When I visit my dashboard ("/dashboard")
If any users have pending orders containing items I sell
Then I see a list of these orders.
Each order listed includes the following information:
- the ID of the order, which is a link to the order show page ("/dashboard/orders/15")
- the date the order was made
- the total quantity of my items in the order
- the total value of my items for that order
```

### Checklist:

- [ ] base tests are written
- [ ] base code is written to pass tests
- [ ] edge case ("sad path") tests written (if required)
- [ ] edge case code written (if required)
- [ ] code is reviewed
- [ ] branch is merged
- [ ] all tests still pass

child of #33

child of #79

depends on #32

### LINKED ISSUE #32: User Story 39, Merchant Dashboard Show Page
```
As a merchant user
When I visit my dashboard ("/dashboard")
I see my profile data, but cannot edit it
```

### Checklist:

- [ ] base tests are written
- [ ] base code is written to pass tests
- [ ] edge case ("sad path") tests written (if required)
- [ ] edge case code written (if required)
- [ ] code is reviewed
- [ ] branch is merged
- [ ] all tests still pass

child of #33

child of #79

depends on #70

depends on #74

### LINKED ISSUE #39: User Story 33, User Profile displays Orders
```
As a registered user
When I visit my Profile Orders page, "/profile/orders"
I see every order I've made, which includes the following information:
- the ID of the order, which is a link to the order show page
- the date the order was made
- the date the order was last updated
- the current status of the order
- the total quantity of items in the order
- the grand total of all items for that order
```

### Checklist:

- [ ] base tests are written
- [ ] base code is written to pass tests
- [ ] edge case ("sad path") tests written (if required)
- [ ] edge case code written (if required)
- [ ] code is reviewed
- [ ] branch is merged
- [ ] all tests still pass

child of #41

child of #79

depends on #57

### LINKED ISSUE #40: User Story 32, User Profile displays Orders link
```
As a registered user
When I visit my Profile page
And I have orders placed in the system
Then I see a link to my Profile Orders page

This link is also in the nav bar, but this new link is for admins.
```

### Checklist:

- [ ] base tests are written
- [ ] base code is written to pass tests
- [ ] edge case ("sad path") tests written (if required)
- [ ] edge case code written (if required)
- [ ] code is reviewed
- [ ] branch is merged
- [ ] all tests still pass

child of #41

child of #79

depends on #57

### LINKED ISSUE #41: Epic: User Order Show Page
## User Order Show Page
The show page template for an order can be shared between users, merchants and admins to DRY up our presentation logic. They will operate through separate controllers, though.

### User Control
- Users can cancel an order if at least one item in the order is NOT yet fulfilled
- When an order is cancelled, any fulfilled items have their inventory returned to their respective merchants

### Merchant Control
- Merchants only see items in the order that are sold by that merchant
- Items from other merchants are hidden

### Admin Control
- Admins can cancel an order on behalf of a user
- Admins can fulfill items on order on behalf of a merchant

### LINKED ISSUE #42: User Story 31, Registered users can check out
```
As a registered user
When I add items to my cart
And I visit my cart
I see a button or link indicating that I can check out
And I click the button or link to check out
An order is created in the system, which has a status of "pending"
I am taken to my orders page ("/profile")
I see a flash message telling me my order was created
I see my new order listed on my profile orders page
My cart is now empty
```

### Checklist:

- [ ] base tests are written
- [ ] base code is written to pass tests
- [ ] edge case ("sad path") tests written (if required)
- [ ] edge case code written (if required)
- [ ] code is reviewed
- [ ] branch is merged
- [ ] all tests still pass

child of #49

child of #79

depends on #47

### PR DESCRIPTION
**Pull request by csvlewis**
- Wrote user methods for selecting a merchant’s total items, total amount, and pending orders
- Wrote order methods for selecting a merchant’s top items, items sold by quantity, items sold by percentage, top states, top cities, top customer by orders, top customer by items, and top spenders
- Created unstyled view for merchant dashboard
- Created route for merchant dashboard orders
- Wrote feature and model tests for all functionality in User Stories 39, 40, 41, and 42
- Closes #29, closes #30, closes #31, closes #32

The organization on the merchant dashboard view is not great at the moment but I was able to make all the tests pass for now at least. I had to do some pretty ugly workarounds to make a couple tests simulate the view that should be displayed, but I figured we could fix those when we get to more styling.

### CODE REVIEW

File: app/controllers/merchant/items_controller.rb [hunk 1]
```diff
@@ -1,6 +1,6 @@
 class Merchant::ItemsController < Merchant::BaseController
   def index
-    @items = current_user.items
+    @item = Item.merchant_items(current_user)
   end
```
Comment (timnallen) [inline app/controllers/merchant/items_controller.rb]: For this one... is @item a group of items? Shouldn't we stick with the plural?

**Review by timnallen — APPROVED** [review]
Really great job with all this. Very productive day.

**Review by allieIsOkay — APPROVED** [review]
Looks good, thanks Chris.

File: app/controllers/merchant/items_controller.rb (hunk 1 above)
Comment (csvlewis) [reply app/controllers/merchant/items_controller.rb]: You're right, I think I just missed that when I rewrote it. I'll make the change right now.

Comment (csvlewis) [reply app/controllers/merchant/items_controller.rb]: Fixed it. Thanks, Tim!

Comment (timnallen) [reply app/controllers/merchant/items_controller.rb]: Thanks Chris!

File: app/models/item.rb [hunk 2]
```diff
@@ -59,6 +59,10 @@ def average_fulfillment_time
     end
   end
 
+  def self.merchant_items(merchant)
```
Comment (allieIsOkay) [inline app/models/item.rb]: Wouldn't this be the same as calling merchant.items?

File: app/controllers/merchant/items_controller.rb (hunk 1 above)
Comment (allieIsOkay) [reply app/controllers/merchant/items_controller.rb]: I'm a little curious why we changes from the current_user.items to this new method. Sorry if i'm not getting something but the original seems more straightforward.

File: config/routes.rb [hunk 3]
```diff
@@ -11,6 +11,8 @@
   get '/login', to: 'sessions#new'
   post '/login', to: 'sessions#create'
   get '/logout', to: 'sessions#destroy'
+  get '/dashboard', to: 'merchant/users#show'
```
Comment (allieIsOkay) [inline config/routes.rb]: Don't we have these two paths from the scope operation on line 19?

**Review by allieIsOkay — COMMENTED** [review]
Just wanted to take another pass over this and had a few questions. Everything still looks good, just wanted to double check on these items.

File: app/models/item.rb (hunk 2 above)
Comment (timnallen) [reply app/models/item.rb]: Hmm... Yes, I think you're right.

File: config/routes.rb (hunk 3 above)
Comment (timnallen) [reply config/routes.rb]: I think you're right here, too. I love these discoveries as they do not require us to do anything but delete something we don't need.

File: app/controllers/merchant/items_controller.rb (hunk 1 above)
Comment (csvlewis) [reply app/controllers/merchant/items_controller.rb]: I'm not sure if the current_user.items method was here when I wrote this but you're right Peter, it is easier to just access it through the current user's items. I think I was stuck in ActiveRecord mode when I wrote this and forgot that there was an easier way. I'll switch it back and remove the unnecessary model test for this method.

File: app/models/item.rb (hunk 2 above)
Comment (csvlewis) [reply app/models/item.rb]: Yep, I'm removing this method from the pull request now to clean things up. Thanks for the suggestion!

File: config/routes.rb (hunk 3 above)
Comment (csvlewis) [reply config/routes.rb]: You're right again, Peter. I don't think I had up to date routes on this branch so I wrote these in manually.

**Review by teresa-m-knowles — APPROVED** [review]

**Review by teresa-m-knowles — APPROVED** [review]
Thanks for the changes, looks good
